volatility3/framework/symbols/linux/utilities/modules.py

```diff
@@ -1,77 +1,122 @@
 
-    @classmethod
+    class ModuleInfo(NamedTuple):
+        """
+        Used to track the name and boundary of a kernel module
+        """
+
+        offset: int
+        name: str
+        start: int
+        end: int
+
+    @staticmethod
     def module_lookup_by_address(
-        cls,
         context: interfaces.context.ContextInterface,
-        layer_name: str,
-        modules: Iterable[extensions.module],
+        kernel_module_name: str,
+        modules: Iterable[ModuleInfo],
         target_address: int,
-    ) -> Optional[extensions.module]:
+        run_hidden_modules: bool = True,
+    ) -> Optional[Tuple[ModuleInfo, Optional[str]]]:
         """
         Determine if a target address lies in a module memory space.
         Returns the module where the provided address lies.
 
+        `modules` must contain masked addresses via `get_module_info_for_module` or
+        a ValueError will be thrown
+
         Args:
             context: The context on which to operate
             layer_name: The name of the layer on which to operate
             modules: An iterable containing the modules to match the address against
             target_address: The address to check for a match
 
         Returns:
-            The first memory module in which the address fits
+            The first memory module in which the address fits and the symbol name for `target_address`
 
         Kernel documentation:
             "within_module" and "within_module_mem_type" functions
         """
+        kernel = context.modules[kernel_module_name]
+
+        kernel_layer = context.layers[kernel.layer_name]
+
+        if modules[0].start != modules[0].start & kernel_layer.address_mask:
+            raise ValueError(
+                "Modules list must be gathered from `run_modules_scanners` to be used in this function"
+            )
+
         matches = []
-        seen_addresses = set()
         for module in modules:
-            _, start, end = cls.mask_mods_list(context, layer_name, [module])[0]
-            if (
-                start <= target_address < end
-                and module.vol.offset not in seen_addresses
-            ):
+            if module.start <= target_address < module.end:
                 matches.append(module)
-                seen_addresses.add(module.vol.offset)
 
-        if len(matches) > 1:
-            warnings.warn(
-                f"Address {hex(target_address)} fits in modules at {[hex(module.vol.offset) for module in matches]}, indicating potential modules memory space overlap.",
-                UserWarning,
-            )
-            return matches[0]
-        elif len(matches) == 1:
-            return matches[0]
+        if len(matches) >= 1:
+            if len(matches) > 1:
+                warnings.warn(
+                    f"Address {hex(target_address)} fits in modules at {[hex(module.start) for module in matches]}, indicating potential modules memory space overlap.",
```

Comment:
Since this doesn't then bomb out, you should say which one will win.

Reply:
added